### ISSUE TITLE
gridsearch method fails with template models

### ISSUE THREAD

**Issue by nplee** [sherpa/sherpa#309]
While using the 11/15 CIAOD build, a name error is thrown when a template model is fitted using the "gridsearch" method's default setting where no sequence is set in Python 3.  In Python 2, there's no problem.

From the "template_model" directory in the tarball, a failure is thrown:

```
sherpa-1> load_ascii('source.dat', ncols=3, dstype=Data1D)
sherpa-2> from sherpa.models import KNNInterpolator
sherpa-3> load_template_model("kerr_templ", "templ_index.dat", KNNInterpolator)
sherpa-4> set_model(kerr_templ)
sherpa-5> set_method("gridsearch")
sherpa-6> set_method_opt("sequence",None)
sherpa-7> fit()
NameError: name 'xrange' is not defined
```

If the "sequence" option is set, the fit will run in both Python versions.

```
sherpa-7> set_method_opt('sequence', kerr_templ.parvals)
sherpa-8> fit()
Dataset               = 1
Method                = gridsearch
Statistic             = chi2
Initial fit statistic = 1.92175e+08
Final fit statistic   = 7.60754e+07 at function evaluation 106
Data points           = 46
Degrees of freedom    = 43
Probability [Q-value] = 0
Reduced statistic     = 1.7692e+06
Change in statistic   = 1.161e+08
   kerr_templ.mass   10          
   kerr_templ.rate   0.3         
   kerr_templ.angle   0.75        
WARNING: parameter value kerr_templ.mass is at its maximum boundary 10.0
```

~Nick

[template_model.tar.gz](https://github.com/sherpa/sherpa/files/593246/template_model.tar.gz)

**Comment by DougBurke** [comment]
Using the master branch, the failure is thrown in the gridsearch code (so I don't think we need the template model):

```
In [1]: from sherpa.astro.ui import *
WARNING: failed to import sherpa.plot.pylab_backend; plotting routines will not be available
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: load_ascii('source.dat', ncols=3, dstype=Data1D)

In [3]: from sherpa.models import KNNInterpolator

In [4]: load_template_model("kerr_templ", "templ_index.dat", KNNInterpolator)

In [5]:  set_model(kerr_templ)

In [6]: set_method("gridsearch")

In [7]: set_method_opt("sequence",None)

In [8]: fit()
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-8-749b6e6de45a> in <module>()
----> 1 fit()

<string> in fit(id, *otherids, **kwargs)

/home/djburke/sherpa/sherpa-base/sherpa/astro/ui/utils.py in fit(self, id, *otherids, **kwargs)
   9301         """
   9302         kwargs['bkg_only'] = False
-> 9303         self._fit(id, *otherids, **kwargs)
   9304 
   9305     def fit_bkg(self, id=None, *otherids, **kwargs):

/home/djburke/sherpa/sherpa-base/sherpa/astro/ui/utils.py in _fit(self, id, *otherids, **kwargs)
   9405                     i).mask & numpy.isfinite(self.get_data(i).get_x())
   9406 
-> 9407         res = f.fit(**kwargs)
   9408         res.datasets = ids
   9409         self._fit_results = res

/home/djburke/sherpa/sherpa-base/sherpa/fit.py in run(fit, *args, **kwargs)
     49     def run(fit, *args, **kwargs):
     50         fit.model.startup()
---> 51         result = func(fit, *args, **kwargs)
     52         fit.model.teardown()
     53         return result

/home/djburke/sherpa/sherpa-base/sherpa/fit.py in fit(self, outfile, clobber)
   1084                                    self.model.thawedpars,
   1085                                    self.model.thawedparmins,
-> 1086                                    self.model.thawedparmaxes)
   1087         # LevMar always calculate chisquare, so call calc_stat
   1088         # just in case statistics is something other then chisquare

/home/djburke/sherpa/sherpa-base/sherpa/fit.py in fit(self, statfunc, pars, parmins, parmaxes, statargs, statkwargs)
    884         if not self.iterate:
    885             return self.method.fit(statfunc, pars, parmins, parmaxes,
--> 886                                    statargs, statkwargs)
    887         else:
    888             return self.current_func(statfunc, pars, parmins, parmaxes,

/home/djburke/sherpa/sherpa-base/sherpa/optmethods/__init__.py in fit(self, statfunc, pars, parmins, parmaxes, statargs, statkwargs)
     95             return statfunc(pars, *statargs, **statkwargs)
     96 
---> 97         output = self._optfunc(cb, pars, parmins, parmaxes, **self.config)
     98 
     99         success = output[0]

/home/djburke/sherpa/sherpa-base/sherpa/optmethods/optfcts.py in grid_search(fcn, x0, xmin, xmax, num, sequence, numcores, maxfev, ftol, method, verbose)
    313     if sequence is None:
    314         ranges = []
--> 315         for index in xrange( npar ):
    316             ranges.append( [ xmin[ index ], xmax[ index ] ] )
    317         sequence = make_sequence( ranges, num )

NameError: name 'xrange' is not defined

In [9]: 
```

**Comment by DougBurke** [comment]
It looks like we have examples of the gridsearch optimiser for the template model code, so we can re-use these to create a test that triggers this:

```
% cd sherpa-test-data/sherpatest/
% python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sherpa import ui
WARNING: failed to import sherpa.plot.pylab_backend; plotting routines will not be available
>>> ui.load_data("default_interp", "load_template_with_interpolation-bb_data.dat")
>>> ui.load_template_model('bb1', "bb_index.dat")
>>> ui.set_model("default_interp", bb1)
>>> ui.set_method('gridsearch')
>>> ui.set_method_opt('sequence', None)  # should be default, but set explicitly
>>> ui.calc_stat()
3.326193401840517e-16
>>> ui.fit()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in fit
  File "/home/djburke/sherpa/sherpa-base/sherpa/ui/utils.py", line 7981, in fit
    res = f.fit(**kwargs)
  File "/home/djburke/sherpa/sherpa-base/sherpa/fit.py", line 51, in run
    result = func(fit, *args, **kwargs)
  File "/home/djburke/sherpa/sherpa-base/sherpa/fit.py", line 1086, in fit
    self.model.thawedparmaxes)
  File "/home/djburke/sherpa/sherpa-base/sherpa/fit.py", line 886, in fit
    statargs, statkwargs)
  File "/home/djburke/sherpa/sherpa-base/sherpa/optmethods/__init__.py", line 97, in fit
    output = self._optfunc(cb, pars, parmins, parmaxes, **self.config)
  File "/home/djburke/sherpa/sherpa-base/sherpa/optmethods/optfcts.py", line 315, in grid_search
    for index in xrange( npar ):
NameError: name 'xrange' is not defined
>>> 
``` 

If I try this with CIAO 4.8 then the fit succeeds and gives me the following (but I don't know if it's worth actually checking the results in this case, since the statistic value is essentially 0 - i.e. perhaps the test should just check that fit does not fail or error out at this time):

```
>>> ui.calc_stat()
3.326193401840517e-16
>>> ui.fit()
Dataset               = default_interp
Method                = gridsearch
Statistic             = chi2gehrels
Initial fit statistic = 3.32619e-16
Final fit statistic   = 9.04592e-18 at function evaluation 257
Data points           = 100
Degrees of freedom    = 98
Probability [Q-value] = 1
Reduced statistic     = 9.23053e-20
Change in statistic   = 3.23573e-16
   bb1.t          2950        
   bb1.dummy      0           
WARNING: parameter value bb1.t is at its maximum boundary 2950.0
WARNING: parameter value bb1.dummy is at its minimum boundary 0.0
WARNING: parameter value bb1.dummy is at its maximum boundary 0.0
>>> 
```

**Comment by DougBurke** [comment]
I have a quick fix for this in #310.
